LX/src/main/java/heronarts/lx/LXSnapshotController.java

```diff
@@ -0,0 +1,161 @@
+package heronarts.lx;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.TreeMap;
+
+import com.google.gson.JsonArray;
+import com.google.gson.JsonElement;
+import com.google.gson.JsonObject;
+
+import heronarts.lx.osc.OscMessage;
+
+public class LXSnapshotController extends LXComponent {
+
+  private Integer nextShapshotID = 1;
+
+  private final List<LXSnapshot> snapshots = new ArrayList<LXSnapshot>();
+  private final TreeMap<Integer, LXSnapshot> snapshotsDict = new TreeMap<Integer, LXSnapshot>();
+
+  public interface SnapshotListener {
+    public void snapshotAdded(LXSnapshotController controller, LXSnapshot snapshot);
+    public void snapshotRemoved(LXSnapshotController controller, LXSnapshot snapshot);
+  }
+
+  private final List<SnapshotListener> snapshotListeners = new ArrayList<SnapshotListener>();
+
+  public LXSnapshotController(LX lx) {
+    super(lx);
+    this.label.setDescription("Snapshots");
+  }
+
+  private LXSnapshot _createNewSnapshot() {
+    LXSnapshot ss = new LXSnapshot(this.lx, this.nextShapshotID);
+    ss.initialize();
+    this.nextShapshotID++;
+
+    return ss;
+  }
+
+  private LXSnapshot _addSnapshot(LXSnapshot snapshot) {
+    if (snapshot == null) {
+      throw new IllegalArgumentException("Cannot add null snapshot");
+    }
+    this.snapshots.add(snapshot);
+    this.snapshotsDict.put(snapshot.snapshotID, snapshot);
+    for (SnapshotListener snapshotListener : this.snapshotListeners) {
+      snapshotListener.snapshotAdded(this, snapshot);
+    }
+    return snapshot;
+  }
+
+  public LXSnapshot addSnapshot(LXSnapshot snapshot) {
+    return _addSnapshot(snapshot);
+  }
+
+  public LXSnapshot addSnapshot() {
+    LXSnapshot snapshot = _createNewSnapshot();
+    _addSnapshot(snapshot);
+    return snapshot;
+  }
+
+  public void runSnapshot(int snapshotID) {
+    if (snapshotsDict.containsKey(snapshotID)) {
+      snapshotsDict.get(snapshotID).run();
+    } else {
+      System.err.println("Unknown snapshotID, cannot run snapshot.");
+    }
+  }
+
+  public LXSnapshotController removeSnapshot(LXSnapshot snapshot) {
+    this.snapshotsDict.remove(snapshot.snapshotID);
+    this.snapshots.remove(snapshot);
+    for (SnapshotListener snapshotListener : this.snapshotListeners) {
+      snapshotListener.snapshotRemoved(this, snapshot);
+    }
+    snapshot.dispose();
+
+    if (this.snapshots.size()==0) {
```

Comment:
I think it's probably better to always reindex the snapshots here for consistency. Maintaining gaps in the indices seems likely to cause confusion. See removeModulator() on LXModulatorComponent

Reply:
I think snapshotIDs (or snapshot indexes) need to remain constant once they're created, because I'm setting up OSC links to them from an external program.  I.E., "load snapshot 7 when I press this button" by sending OSC command \lx\snapshot\7.  Wouldn't want to reindex for instance if I delete snapshot 2 because then my external program is out of sync.  My reasoning was that the one time it is acceptable to start the snapshotIDs/indexes over is if the entire list is cleared.  Then it's nice to reset to 1 to keep things clean.  This was part of my reason for using the name snapshotID instead of index, because index implies sequential ordering.  The function here is a unique identifier that remains constant with the snapshot instance.